lib/common/zstdmt_priority_threadpool.h

```diff
@@ -0,0 +1,59 @@
+#ifndef ZSTDMT_PRIORITY_THREADPOOL_H_MODULE
+#define ZSTDMT_PRIORITY_THREADPOOL_H_MODULE
+
+#include <stddef.h>
+
+/* Priority thread pool context */
+typedef struct ZSTDMT_pThreadPoolCtx_s ZSTDMT_pThreadPoolCtx_t;
+
+/* Callback function for each job */
+typedef void (*ZSTDMT_pThreadPool_fn)(void* data);
+
+/* Basic Usage:
+ * ctx = ZSTDMT_pThreadPoolCtx_create(QUEUE_SIZE, NB_THREADS);
+ * for (i = 0; i < nbJobs; ++i)
+ *    ZSTDMT_pThreadPool_addJob(ctx, priority(i), 0, callbackFn, NULL);
+ * ZSTDMT_pThreadPool_waitAllThreads(ctx);
+ * ZSTDMT_pThreadPool_free(ctx); */
+/***************************************
+* ZSTDMT_pThreadPool API
+***************************************/
+
+/* Create a priority thread pool of 'queueSize' using 'nbThreads' worker threads
+ * @return : ptr to ZSTDMT_pThreadPoolCtx_t for thread pool */
+static ZSTDMT_pThreadPoolCtx_t* ZSTDMT_pThreadPoolCtx_create(size_t queueSize,
+  size_t nbThreads);
+
+/* Destroies the context */
+static size_t ZSTDMT_pThradPoolCtx_free(ZSTDMT_pThreadPoolCtx_t* ctx);
+
+/* Enqueue a job into the priority queue of the thread pool with given 'priority'
+ * Note: time complexity will be log(#jobs)
+ * Note: lower priorities are going to be excuted first (ie. priority 0 jobs will come first)
+ * @return : ID of the job */
+static size_t ZSTDMT_pThreadPool_addJob(ZSTDMT_pThreadPoolCtx_t* ctx, size_t priority,
+  size_t addedTimeMs, ZSTDMT_pThreadPool_fn fn, void* data);
+
+/* Dequeue the job from the priority queue of the thread pool if it exists in the queue
+ * @return : 1 for success or some error code */
+static size_t ZSTDMT_pThreadPool_removeJob(ZSTDMT_pThreadPoolCtx_t* ctx, size_t jobId);
```

Comment:
We will have to go to the specifics of the algorithm that makes you need a `removeJob` action.
I see several actions below related to some form of "manual" maintenance of the queue,
it's important to study if that's the best way to do it.

Reply:
I don't see the multi-threaded decoder algorithm using this. Just thought it would be useful to support for future use cases. Perhaps this would be better left to when such a use case comes about?

Reply:
Yes, probably.
Without a clear use case, we can't be sure that this functionality is useful, nor if it's the right abstraction.
Try to reduce the API to what you will need.